File: driver-dna/backend/routes/driver.py
"""Driver-related API routes."""

from fastapi import APIRouter, Query

from schemas import DriverMetricsResponse, DriversResponse
from services.driver_service import get_driver_metrics, list_drivers

router = APIRouter()


@router.get("/drivers", response_model=DriversResponse)
def get_drivers(
    year: int = Query(..., description="Season year"),
    force_refresh: bool = Query(False, description="Bypass memory cache"),
) -> DriversResponse:
    drivers, cached = list_drivers(year=year, force_refresh=force_refresh)
    return DriversResponse(year=year, drivers=drivers, cached=cached)


@router.get("/driver/{code}", response_model=DriverMetricsResponse)
def get_driver(
    code: str,
    year: int = Query(..., description="Season year"),
    force_refresh: bool = Query(False, description="Bypass memory cache"),
) -> DriverMetricsResponse:
    payload, cached = get_driver_metrics(code=code, year=year, force_refresh=force_refresh)
    payload = {**payload, "cached": cached}
    return DriverMetricsResponse(**payload)

File: driver-dna/frontend/src/api/driver.ts
import { apiGet } from "./client";
import type { DriverMetricsResponse, DriversResponse } from "../types/api";

const DRIVER_PAGE_QUERY_PROFILE = {
  max_rounds: 4,
  max_telemetry_laps: 3
} as const;

export const getDriversByYear = (
  year: number,
  signal?: AbortSignal
): Promise<DriversResponse> => {
  return apiGet<DriversResponse>("/api/drivers", { year }, signal);
};

export const getDriverMetrics = (
  year: number,
  code: string,
  signal?: AbortSignal
): Promise<DriverMetricsResponse> => {
  return apiGet<DriverMetricsResponse>(
    `/api/driver/${code.toUpperCase()}`,
    { year, ...DRIVER_PAGE_QUERY_PROFILE },
    signal
  );
};
